### DISCUSSION TITLE
Incremental search

### DISCUSSION THREAD

**Discussion by fbergroth** [meow-edit/meow#175] [Ideas]
I think adding incremental search would be nice.

There are some issues with entering a search pattern via `meow-visit` if `meow-visit-sanitize-completion=t`.

- The pattern is always quoted, so searching for a regex will fail (but docstring says it'll do a regex search!)
- If you have a buffer containing `meow` and `meow-variable`, you can't easily search for the prefix `meow`, since symbol boundaries will be added via completions (typing `meow` results in `\_<meow\_>`) 
- No incremental feedback

I'm experimenting with binding `meow-isearch` to `/`, which I quite like.

```elisp
(defun meow-isearch ()
  (interactive)
  (if (meow--direction-backward-p) ; not sure how should work
      (isearch-backward t)
    (isearch-forward t))
  (when isearch-success
    (thread-first
      (meow--make-selection '(select . visit) isearch-other-end (point))
      (meow--select))))
```

Now, there is some overlap between this and `meow-visit`. I think it would be nice if `v` did `meow-visit` with completions being words (`\<word\>`), and `V` did `meow-visit` with completions being symbols.

What do you think @DogLooksGood @eshrh?

**Comment by eshrh** [comment]
I've encountered some of the problems you mentioned (not being able to easily search for a part of another symbol is a big one, missing regex is another), so personally, I actually have `v` bound to `swiper`. I think binding an isearch on `/` by default is a good idea, since that key is currently unoccupied.

To check if you're going backward I think you can check whether the variable `current-prefix-arg` is `string=` to `"-"`. The `-` key in meow just modifies the prefix arg by simulating `C-u`.

As for meow-visit, I am totally in favor of binding v and V to word and symbol. This is in-line with other current keybinds and seems like a very useful addition that saves a lot of C-n presses when looking for just words.

**Comment by DogLooksGood** [comment]
I will not add incremental searching for `meow-visit`, because `meow-visit` use the builtin `completing-read`, it can be baked by a lot packages, like helm, ivy, vertico, selectrum, etc. 

But I can definitely understand what you pointed out.  I'd like to have an integration with incremental search, but not by providing a command. Because there are builtins.  Additionally, there are some packages provides the enhance alternatives, like phi-search and ctrlf.  I haven't do it because I have no idea how to get a general solution.

For isearch, I'd like to have something like this
```
(defun meow--post-isearch-function ()
  (unless isearch-mode-end-hook-quit
    (when isearch-success
      (let ((beg (car isearch-match-data))
	    (end (cadr isearch-match-data)))
	
	(thread-first
	  (meow--make-selection '(select . visit)
				beg
				(if isearch-forward end isearch-other-end))
	  (meow--select (not isearch-forward)))))))

(add-hook 'isearch-mode-end-hook 'meow--post-isearch-function)
```

It should work for both `C-s`, `C-r`, `C-M-s`, but it doesn't work for other packages, obviously.
I'll add incremental search support, whenever I found nice way to integrate.

**Reply by fbergroth** [reply]
> I will not add incremental searching for `meow-visit`, because `meow-visit` use the builtin `completing-read`, it can be baked by a lot packages, like helm, ivy, vertico, selectrum, etc.

I didn't suggest this. I suggested to use words/symbols as completions for v/V once you have incremental search bound to some other key.

> For isearch, I'd like to have something like this [...]

Looks good. It's indeed cleaner to handle selection integration this way.

> I'd like to have an integration with incremental search, but not by providing a command. Because there are builtins.

It's not needed, but I think it does make sense to search forward on `/`, and search backward with negative argument, to make it more in-line with meow.
```
(defun meow-isearch (arg)
  (interactive "P")
  (if (eq arg '-)
      (isearch-backward t)
    (isearch-forward t)))
```

**Reply by DogLooksGood** [reply]
You can't do 
```
(if (meow--with-negative-argument-p arg)
    (isearch-backward t)
  (isearch-forward t))
;; then create selection here
```
You have to create selection after user finish searching.  However `isearch-mode-end-hook` is not respected by other packages.
So, at the moment, user either write their wrapper commands, or use the integration like I post above.

**Reply by fbergroth** [reply]
> It should work for both `C-s`, `C-r`, `C-M-s`, but it doesn't work for other packages, obviously.
> I'll add incremental search support, whenever I found nice way to integrate.

Meanwhile, can we add `meow--post-isearch-function` to meow-shims? It fixes that `meow-search` will re-use the pattern from isearch.

**Reply by DogLooksGood** [reply]
I think a variable with function is not enough here. Implementing `meow-search` with  `isearch`, `phi-isearch` and `ctrlf` are quite different.

**Reply by fbergroth** [reply]
It's not what I mean. If I do `C-s some-pattern RET`, I expect `M-x meow-search` to search forward using `some-pattern`. Currently it does not, but it works with `meow--post-isearch-function`.

**Reply by DogLooksGood** [reply]
`meow-search` will use `regexp-search-ring`. 
Currently, `meow-mark-word`, `meow-mark-symbol` and `meow-visit` will push regexp into this variable.

You want provide a custom function to control which variable to be used?

**Reply by fbergroth** [reply]
Ah sorry. `C-s some-pattern RET` activated a region, and then I called `meow-search`, which pushed it into `regexp-search-ring`. I thought `meow--post-isearch-function` was inserting the pattern into the ring. :man_facepalming:

**Reply by DogLooksGood** [reply]
It will be better if a region is activated after isearch, and cancel the existing region before isearch.

But changing the behavior of builtin commands is not good. So I'd like to make no changes.
